### PR TITLE
Feature/voicemanager

### PR DESCRIPTION
**Pull request by jmidt**
Redesign of the entire voicemanager. 
One bug remains: In interval mode, hitting a key with a voice already playing (from the interval created by a different key) results in a segfault. 

Note that a slight expansion of NanoCanvas has been made. This could be moved somewhere else if that is better.

### CODE REVIEW

**Comment by topisani** [comment]
merge develop into this branch first though, to make sure travis is happy

File: src/core/voices/settings_screen.cpp [hunk 1]
```diff
@@ -50,7 +50,8 @@ namespace otto::core::voices {
     case Encoder::yellow: props.portamento.step(ev.steps); break;
     case Encoder::red: {
       // Fuck it. I can do binary logic myself.
```
[the remark this answers is not in the document]
Comment (jmidt) [reply src/core/voices/settings_screen.cpp]: Okay okay. It took me two tries, but all's well that ends well.

File: src/core/voices/voice_manager.inl [hunk 2]
```diff
@@ -295,6 +461,18 @@ namespace otto::core::voices {
...
       case PlayMode::unison: return "unison";
+      case PlayMode::interval: return "interval";
+      };
+      return "";
+    }
+
+    inline std::string aux_setting(PlayMode pm) noexcept
+    {
+      switch (pm) {
+        case PlayMode::poly: return "M.U.S.";
```
[the remark this answers is not in the document]
Comment (jmidt) [reply src/core/voices/voice_manager.inl]: Or maybe "rand" or something. But yeah, "M.U.S." looks like something out of Fallout

File: src/core/voices/voice_manager.hpp [hunk 3]
```diff
@@ -263,12 +296,52 @@ namespace otto::core::voices {
...
+    };
+
+
     float pitch_bend_ = 1;
 
     props::Property<bool> sustain_ = {false};
 
     std::deque<Voice*> free_voices;
-    std::vector<NoteVoicePair> note_stack;
+    std::deque<NoteVoicePair> note_stack;
```
[the remark this answers is not in the document]
Comment (jmidt) [reply src/core/voices/voice_manager.hpp]: It has to do with the way mono and unison modes work. At the end of the day it boils down to the fact that remove-erase can't use a reverse pointer. So i need push_front a few times.

File: src/core/voices/voice_manager.inl [hunk 4]
```diff
@@ -85,12 +90,231 @@ namespace otto::core::voices {
...
+    auto it = std::remove_if(vm.note_stack.begin(), vm.note_stack.end(), [&](auto nvp) {
+      if (nvp.key == key) {
+        if (nvp.has_voice()) {
+          free_voice(*nvp.voice);
+        }
+        return true;
+      }
+      return false;
+    });
+    vm.note_stack.erase(it, vm.note_stack.end());
```
[the remark this answers is not in the document]
Comment (jmidt) [reply src/core/voices/voice_manager.inl]: word.

File: src/core/voices/voice_manager.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (topisani) [reply src/core/voices/voice_manager.hpp]: you can, you just use the reverse_iterator.base() function to get the normal iterator from the reverse iterator

[the remark this answers is not in the document]
Comment (jmidt) [reply src/core/voices/voice_manager.hpp]: Good to know (although i think it needs to be incremented before calling base()? ).
In any case, is there a good reason to change it? Other than a preference for simpler containers? :)

File: src/core/voices/voice_manager.hpp [hunk 5]
```diff
@@ -100,16 +105,18 @@ namespace otto::core::voices {
     template<typename T, int N>
     friend struct VoiceManager;
 
-    void trigger(int midi_note, float velocity) noexcept;
+    void trigger(int midi_note, float mult, float velocity, bool legato, bool jump) noexcept;
     void release() noexcept;
+    void releaseNoEnv() noexcept;
```
[the remark this answers is not in the document]
Comment (jmidt) [reply src/core/voices/voice_manager.hpp]: Done.

File: scripts/otto_test_visualization.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import sys
```
[the remark this answers is not in the document]
Comment (jmidt) [reply scripts/otto_test_visualization.py]: Removed.

File: src/core/voices/voice_manager.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmidt) [reply src/core/voices/voice_manager.hpp]: Kept for now

**Comment by jmidt** [comment]
I will merge if there are no objections. I discovered a bug, though. With BETTER_ENUM, the props::wrap mixin does not work anymore. It crashes when it tries to wrap a better_enum property.

**Comment by topisani** [comment]
did you replace it with a non-BETTER_ENUM enum on this branch?

**Comment by jmidt** [comment]
Nope. It is also a problem elsewhere. See Arp, for instance. We could change them all to boring old enums?

**Pull request merged**